code/harness/Tester.cpp

```diff
@@ -760,65 +761,121 @@ int Tester::log_snapshot_save(string log_file) {
+    bytes_done += new_amount;
+  }
+
+  fsync(log_fd);
   return SUCCESS;
 }
 
 int Tester::log_snapshot_load(string log_file) {
   // TODO(ashmrtn): What happens if this fails?
+  // TODO(ashmrtn): Change device_clone to be an mmap of the disk we need to get
+  // stuff on.
   int res = ioctl(cow_brd_fd, COW_BRD_WIPE);
   if (res < 0) {
     cerr << "error wiping old disk snapshot" << endl;
   }
-  ifstream log(log_file);
-  char device_clone[device_size];
-  log.read(device_clone, device_size);
-  bool err = log.fail();
-  int errnum = errno;
-  assert(log.peek() == EOF);
-  log.close();
-  if (err) {
-    std::cout << "error " << strerror(errnum) << std::endl;
+
+  // device_size happens to be the number of 1k blocks on cow_brd (from original
+  // brd behavior...), so convert it to a number of bytes.
+  unsigned int dev_bytes = device_size * 2 * 512;
+  unsigned int bytes_done = 0;
+  const unsigned int buf_size = 4096;
+  unsigned int buf[buf_size];
+  int log_fd = open(log_file.c_str(), O_RDONLY);
+  if (log_fd < 0) {
+    cerr << "error opening log file" << endl;
     return LOG_CLONE_ERR;
   }
-  // TODO(ashmrtn): Change device_clone to be an mmap of the disk we need to put
-  // stuff on.
-  int device_fd = open(COW_BRD_PATH, O_WRONLY);
-  res = lseek(device_fd, 0, SEEK_SET);
+
+  // cow_brd_fd is RDONLY.
+  int device_path = open(COW_BRD_PATH, O_WRONLY);
+  if (device_path < 0) {
+    cerr << "error opening log file" << endl;
+    return LOG_CLONE_ERR;
+  }
+
+  res = lseek(device_path, 0, SEEK_SET);
   if (res < 0) {
     cerr << "error seeking to start of test device" << endl;
   }
-  unsigned int bytes_written = 0;
-  do {
-    int res = write(device_fd, (void*) (device_clone + bytes_written),
-        device_size - bytes_written);
-    if (res < 0) {
-      int err = errno;
-      cerr << "error writing snapshot to test device " << errno << endl;
-      cerr << "aborting copy after " << bytes_written << endl;
-    }
-    bytes_written += res;
-  } while (bytes_written < device_size);
-  fsync(device_fd);
-  close(device_fd);
+  res = lseek(log_fd, 0, SEEK_SET);
+  if (res < 0) {
+    cerr << "error seeking to start of test device" << endl;
+  }
+  while (bytes_done < dev_bytes) {
+    // Read a block of data from the base disk image.
+    unsigned int bytes = 0;
+    unsigned int new_amount = (dev_bytes < bytes_done + buf_size)
+                            ? dev_bytes - bytes_done
+                            : buf_size;
+    do {
+      int res = read(log_fd, buf + bytes, new_amount - bytes);
+      if (res < 0) {
+        cerr << "error reading from raw device to log disk snapshot" << endl;
+        break;
+      }
+      bytes += res;
+    } while (bytes < new_amount);
+
+    // Write a block of data to the log file.
+    bytes = 0;
+    do {
+      int res = write(device_path, buf + bytes, new_amount - bytes);
+      if (res < 0) {
+        cerr << "error reading from raw device to log disk snapshot" << endl;
+        break;
```

Comment:
return here? You break if write is not successful, but you still add new_amount to bytes_done